### PR TITLE
msgr2: frame tag protocol validation plus frame classes on its own header

### PR DESCRIPTION
**Pull request by rjfd**
This PR implements the following two changes:
* move frame classes into its own header
* tag validation on read_frame

Signed-off-by: Ricardo Dias <rdias@suse.com>

### CODE REVIEW

File: src/msg/async/ProtocolV2.cc [hunk 1]
```diff
@@ -69,8 +69,9 @@ void ProtocolV2::run_continuation(CtPtr continuation) {
 
 #define READB(L, B, C) read(CONTINUATION(C), L, B)
 
-#ifdef UNIT_TESTS_BUILT
+#define TAG(T) (1 << ((uint64_t)(T)-1))
```
Comment (liewegas) [inline src/msg/async/ProtocolV2.cc]: TAG_MASK to emphasize it's a bitmask?

**Comment by liewegas** [comment]
otherwise this looks reasonable!  i like how this drops the get_buffer() bit everywhere

File: src/msg/async/ProtocolV2.cc (hunk 1 above)
Comment (rjfd) [reply src/msg/async/ProtocolV2.cc]: sure, will change that

**Comment by rjfd** [comment]
@liewegas @rzarzynski I added https://github.com/ceph/ceph/pull/26628/commits/26396a73c3c4daaf998d5ff1ef62f08675bb10da (move frame classes into header) into this PR

**Comment by rjfd** [comment]
@liewegas will need to schedule a new build in shaman and then run tests again. Where is the monthrash change you told me about?

**Comment by liewegas** [comment]
@rjfd https://github.com/ceph/ceph/pull/25897

**Comment by rjfd** [comment]
> nit: could add a reference to doc/dev/msgr2.rst in the new header

done

**Comment by rjfd** [comment]
jenkins retest this please

**Comment by liewegas** [comment]
```
2019-02-26 02:22:47.248 7f7409e20700 -1 --2-  >> [v2:172.21.15.98:3303/0,v1:172.21.15.98:6792/0] conn(0x7f740c1523f0 0x7f740c1505f0 unknown :-1 s=CONNECTING pgs=0 cs=0 l=0 rx=0 tx=0).handle_read_frame_preamble_main received unexpected tag: expected=0x28 got=0x4
2019-02-26 02:22:47.248 7f740a621700 -1 --2-  >> [v2:172.21.15.139:3301/0,v1:172.21.15.139:6790/0] conn(0x7f740c159290 0x7f740c15b730 unknown :-1 s=CONNECTING pgs=0 cs=0 l=0 rx=0 tx=0).handle_read_frame_preamble_main received unexpected tag: expected=0x28 got=0x4
2019-02-26 02:22:47.449 7f7409e20700 -1 --2-  >> [v2:172.21.15.98:3303/0,v1:172.21.15.98:6792/0] conn(0x7f740c1523f0 0x7f740c1505f0 unknown :-1 s=CONNECTING pgs=0 cs=0 l=0 rx=0 tx=0).handle_read_frame_preamble_main received unexpected tag: expected=0x28 got=0x4
2019-02-26 02:22:47.450 7f740a621700 -1 --2-  >> [v2:172.21.15.139:3301/0,v1:172.21.15.139:6790/0] conn(0x7f740c159290 0x7f740c15b730 unknown :-1 s=CONNECTING pgs=0 cs=0 l=0 rx=0 tx=0).handle_read_frame_preamble_main received unexpected tag: expected=0x28 got=0x4
```
/a/sage-2019-02-26_00:43:29-rados-wip-sage-testing-2019-02-25-1642-distro-basic-smithi/3638128

which is just running qa/workunits/mon/caps.sh,
```
expect "ceph auth get-or-create client.bazar > $tmp.bazar.keyring" 0
expect "ceph -k $tmp.bazar.keyring --user bazar mon_status" 13
```
and expecting (but not getting) EACCES

**Comment by rjfd** [comment]
@liewegas I found some problems in the `expected_tags` method in several runs and I've been fixing them. Hope that the push I did one hour ago is the last one.

**Comment by rjfd** [comment]
jenkins retest this please

**Comment by rjfd** [comment]
monthrash suite all green: http://pulpito.ceph.com/rdias-2019-02-26_20:43:06-rados:monthrash-wip-rdias2-testing-distro-basic-smithi/

Running rados suite now.

**Comment by rjfd** [comment]
Rados suite also good: http://pulpito.ceph.com/rdias-2019-02-26_22:35:27-rados-wip-rdias2-testing-distro-basic-smithi/

**Pull request merged**